### PR TITLE
Deep clone transformMatrix, strokeDashArray and styles properties

### LINKED ISSUE #2396: IText Clone object, pass reference to style object instead of cloning it.
1) Add an IText with some initial text eg:-sample, delete all characters(sample) using delete key/backspace key  type new content (eg:-mytext).
    
2) Take clone of edited IText
3) Update color of single character the changes reflected to parent also
    change the color by selecting "my" in mytext

### PR DESCRIPTION
**Pull request by asturur**
This will be hardly noticed by anyone.
But still is a bug.

closes #2396.

### CODE REVIEW

**Comment by asturur** [comment]
Added clone for the iText style.
@kangax what do you think?

**Comment by asturur** [comment]
Added failing tests for the modified properties.
ready to merge for me.

File: src/shapes/itext.class.js [hunk 1]
```diff
@@ -1121,8 +1121,16 @@
      * @return {Object} object representation of an instance
      */
     toObject: function(propertiesToInclude) {
+      var newStyle = { }, i, j, row;
```
Comment (kangax) [inline src/shapes/itext.class.js]: `newStyle` &rarr; `clonedStyles`

File: src/shapes/object.class.js [hunk 2]
```diff
@@ -783,7 +783,7 @@
             fill:                     (this.fill && this.fill.toObject) ? this.fill.toObject() : this.fill,
             stroke:                   (this.stroke && this.stroke.toObject) ? this.stroke.toObject() : this.stroke,
             strokeWidth:              toFixed(this.strokeWidth, NUM_FRACTION_DIGITS),
-            strokeDashArray:          this.strokeDashArray,
+            strokeDashArray:          this.strokeDashArray ? extend([ ], this.strokeDashArray) : this.strokeDashArray,
```
Comment (kangax) [inline src/shapes/object.class.js]: So we're cloning array because... something might modify it somewhere else?

**Comment by asturur** [comment]
if it happens with styles it will happen with array also.

**Comment by kangax** [comment]
But `styles` has objects, array has primitives; can't modify primitives' references

**Comment by asturur** [comment]
I do not understand what  you mean, but the bug is there.
Go to kitchensink, execute this code:

```
// clear canvas
canvas.clear();

// remove currently selected object
canvas.remove(canvas.getActiveObject());

var transform = [2,0,0,2,4,4];
var strokedash = [3, 3];
// add red rectangle
var rect = new fabric.Rect({
  width: 50,
  height: 50,
  left: 50,
  top: 50,
  fill: 'rgb(255,0,0)',
  transformMatrix: transform,
  strokeDashArray: strokedash,
  stroke: 'black',
  strokeWidth: 2
});

// add green, half-transparent circle
canvas.add(rect);
canvas.add(rect.clone().setFill('blue'));
canvas.add(rect.clone().setFill('green'));
```

obtain this
image

then run this code

```
canvas.getActiveObject().transformMatrix[0] = 1;
canvas.getActiveObject().strokeDashArray[1] = 8;
canvas.renderAll();
```

and you get this: ( all object get modified )
image

**Comment by kangax** [comment]
@asturur not sure I understand the bug here... in your last example, you're changing values of transformMatrix and seeing result rendered, no? Isn't this expected behavior?

**Comment by asturur** [comment]
i m changing the transformatrix of one rect, and all three are changing.
because in transformMatrix we have reference of the cloned object instead of a new array.

**Comment by kangax** [comment]
Oh, I didn't realize it's cloned. Of course, you're right.

**Comment by kangax** [comment]
One last thing — we don't want to use extend for arrays since it does for..in iteration and that kind of iteration doesn't just iterate over indices (0..n) but also anything in Array.prototype, etc. Just do transformMatrix.concat() to create a clone of array.

**Comment by asturur** [comment]
@kanga done.

**Comment by kangax** [comment]
+changelog

**Pull request merged**